text/2020-00-00-error-handling.md

```diff
@@ -0,0 +1,297 @@
+
+    #[error(transparent)]
+    Io(#[from] IoError),
+
+    #[error("Invalid transaction tso with start_ts: \
+             {start_ts}, commit_ts: {commit_ts}")]
+    InvalidTxnTso {
+        start_ts: TimeStamp,
+        commit_ts: TimeStamp,
+    },
+
+    #[error(transparent)]
+    Other(#[from] AnyError),
+}
+```
+
+Note that I've elided some variants for concision.
+
+Current pattern (from
+[src/storage/errors.rs](https://github.com/tikv/tikv/blob/dec3d008acb56ad0fb34eeeecca62aaa9091b994/src/storage/errors.rs#L174-L180)
+):
+
+```rust
+    Err(Error(box ErrorInner::Engine(EngineError(box EngineErrorInner::Request(e)))))
+    | Err(Error(box ErrorInner::Txn(TxnError(box TxnErrorInner::Engine(EngineError(
+        box EngineErrorInner::Request(e),
+    ))))))
+    | Err(Error(box ErrorInner::Txn(TxnError(box TxnErrorInner::Mvcc(MvccError(
+        box MvccErrorInner::Engine(EngineError(box EngineErrorInner::Request(e))),
+    )))))) => ...,
+```
+
+Pattern with proposed changes:
+
+```rust
+    Err(Error::Engine(EngineError::Request(e)))
+    | Err(Error::Txn(TxnError::Engine(EngineError::Request(e))))
+    | Err(Error::Txn(TxnError::Mvcc(MvccError::Engine(
+        EngineError::Request(e),
+    )))) => ...,
+```
+
+Once the error migration is complete, we are guaranteed that the second and
+third clauses are impossible, so the code can be further simplified to:
+
+```rust
+    Err(Error::Engine(EngineError::Request(e))) => ...,
+    _ => unreachable!(),
+```
+
+I have performed the complete migration for the error types in components/backup
+and src/storage/txn. The changes can be seen in
+[this branch](https://github.com/tikv/tikv/compare/master...nrc:errors-3?expand=1).
+
+### Changes to thiserror
+
+In my opinion, [thiserror](https://github.com/dtolnay/thiserror) is currently
+the best library for error handling in high-quality code like TiKV. However, for
+TiKV we can do a bit better than the current version by having better support
+for boxed errors and multiple error types.
+
+When a field in an enum variant or struct is marked with `#[from]`, thiserror
+generates an implementation of `From` to convert from an object of the field's
+type to the error type. I propose that this attribute is extended to support an
+optional `boxed` modifer: `#[from(boxed)]`. `#[from(boxed)]` is only valid on a
+field with type `Box<...>`. The `From` impl generated is from the unboxed type
+(i.e., `T` in `Box<T>`) to the error type, rather than the boxed type
+(`Box<T>`).
+
+`#[from(boxed)]` means that bare `?` can be used to convert errors without
+requiring `.map_err(Box::new)`.
+
+My second proposal is for `#[error(unwrap)]` on error types and
+`#[from(unwrap)]` on fields (note that `#[from(boxed, unwrap)]` would be valid).
+Where an error marked with `#[error(unwrap)]` is converted to an error type via
+a `From` impl generated from `#[from(unwrap)]`, then where possible nested
+errors will be unwrapped rather than nested.
```

Comment:
I am not sold on this, or maybe not the prototype implementation of it, or maybe the RFC needs to call out ways this unwrap-conversion goes wrong and conventions for how we plan to prevent them. For example:

- I tried adding an Error::DefinitelyNotRocks(String) variant to tikv::storage::txn::Error and it happily gets converted to backup::Error::Rocks because both are strings. Do we stop using String error variants anymore? What is the general guideline for which types to allow or avoid?

- This forces a model where all errors of the same type go in the same variant, which I see as a step in the wrong direction. In backup::Error I see:

    ```rust
    #[error("IO error {0}")]
    Io(#[from] IoError),
    ```

    which would tend to produce unhelpful messages like "IO error Operation not permitted". I would rather see:

    ```rust
    #[error("failed to create external storage: {0}")]
    CreateStorage(IoError),
    #[error("failed to something else: {0}")]
    SomethingElse(IoError),
    ```

    But that's no longer supported:

    ```console
    error[E0119]: conflicting implementations of trait `thiserror::private::__Rewrap<std::io::Error>` for type `errors::Error`:
      --> components/backup/src/errors.rs:95:17
       |
    95 | #[derive(Debug, Error)]
       |                 ^^^^^
       |                 |
       |                 first implementation here
       |                 conflicting implementation for `errors::Error`
    ```

Together these two mean that type-based unwrap conversion ascribes a global meaning to every error *type*. One error enum can't use `CreateStorage(IoError)` if a different error enum in a different package uses `Io(IoError)` because an unwrap-conversion could end up going from the general to the specific and thus attaching the wrong specific message to a general error, resulting in wrong and misleading user-facing messages.

Do you combat that by declaring that an error's message may be no more specific than its type, i.e. forbid `CreateStorage(IoError)`? That leads to `pub struct CreateStorage(IoError)` and you are back in the world of nesting and repetition: `Error::CreateStorage(CreateStorage(e))`.

It's worth mentioning that the alternative to unwrap-conversions is to just write the From impl which is not so bad:

```rust
impl From<TxnError> for Error {
    fn from(e: TxnError) -> Self {
        match e {
            TxnError::Engine(e) => Error::Engine(e),
            TxnError::Io(e) => Error::Io(e),
            TxnError::Other(e) => Error::Other(e),
            e @ TxnError::Codec(_)
            | e @ TxnError::Mvcc(_)
            | e @ TxnError::ProtoBuf(_)
            | e @ TxnError::InvalidTxnTso { .. }
            | e @ TxnError::InvalidReqRange { .. } => Error::Txn(e),
        }
    }
}
```